server/src/main/java/org/elasticsearch/repositories/IndexMetaDataGenerations.java

```diff
@@ -133,11 +134,18 @@ public IndexMetaDataGenerations withAddedSnapshot(
      * @return new instance without the given snapshot
      */
     public IndexMetaDataGenerations withRemovedSnapshots(Collection<SnapshotId> snapshotIds) {
-        final Map<SnapshotId, Map<IndexId, String>> updatedIndexMetaLookup = new HashMap<>(lookup);
-        updatedIndexMetaLookup.keySet().removeAll(snapshotIds);
-        final Map<String, String> updatedIndexMetaIdentifiers = new HashMap<>(identifiers);
-        updatedIndexMetaIdentifiers.keySet()
-            .removeIf(k -> updatedIndexMetaLookup.values().stream().noneMatch(identifiers -> identifiers.containsValue(k)));
+        final Map<SnapshotId, Map<IndexId, String>> updatedIndexMetaLookup = lookup.entrySet()
+            .stream()
+            .filter(e -> snapshotIds.contains(e.getKey()) == false)
```

Comment:
This is in essence `List.contains`. I wonder if this might be slower than copying it to a Set and then looking up against that? A delete of a big list of snapshots could hurt here maybe?

Reply:
Yeah, I was thinking about fixing it in a separate PR by creating a `Set` in SnapshotsService https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L2420 where `deleteSnapshots` gets called.